### PR TITLE
tests: net: lwm2m: Remove unnecessary filtering

### PR DESCRIPTION
**Pull request by SeppoTakalo**
Content-format tests do not require network or
Newlibc, so by removing these filters I'm able to run tests on qemu_cortex_m3 and native_posix platforms.

### CODE REVIEW

**Comment by SeppoTakalo** [comment]
Added a minor fix into the test code as well, this prevented the build on `qemu_cortex_m3`

**Comment by SeppoTakalo** [comment]
Looks like removing the `depends: netif` triggers way too many platforms that this does not fit, so I added it back.

**Comment by SeppoTakalo** [comment]
Now it seem to build over 700 targets/tests... which does not make sense.

I wonder if I should just limit to `native_posix` and `qemu_cortex_m3` so  I don't need any HW resources, but at least one compiler is targeting Cortex-M type of devices.

**Review by rlubos — DISMISSED** [review]

**Comment by PerMac** [comment]
Adding `platform_allow` means that those tests will be blocked in twister on any platform (for everyone) not listed there. Its usage should be limited to cases, where it is mostly certain that it won't work any where else. It shouldn't be used as a limitation for CI. Having the above in mind, was this the desired effect? I got the impression that the changes are rather to limit CI scope. If so maybe you can consider using `integration_platforms` -> this will specifically tell CI to use those as as scope and also make sure those are not skipped. However, if you remove filters, like depends_on etc. it still will be built during biweekly all-on-all build-tests. There is also another recent feature which might help you to limit the scope: https://github.com/zephyrproject-rtos/zephyr/pull/53110

**Comment by PerMac** [comment]
I played a bit with suggested option and I think combination of platform_key and integration_platforms will do the trick. With platform_key you'll reduce the scope for calls with `--all` like biweekly build or without any scope (when "default" platforms are used). With integration_platforms you'll make sure that the test is executed on you're chosen platform in the on-target CI (when it is scoped due to PR-proposed changes). 
E.g. I used:
```
  net.lwm2m.block_transfer:
    platform_key:
      - simulation
    tags: lwm2m
    integration_platforms:
      - qemu_cortex_m3
      - native_posix
```
and with `--all` the test was build only on 4 platform (although non from integration_platforms and not executed since nsim platforms were selected which I don't have installed), and with -G it was build and executed only on the integration ones

**Review by rlubos — DISMISSED** [review]

**Review by PerMac — DISMISSED** [review]

File: tests/net/lib/lwm2m/content_plain_text/testcase.yaml [hunk 1]
```diff
@@ -1,8 +1,8 @@
-common:
-  depends_on: netif
 tests:
   net.lwm2m.content_plain_text:
-    filter: TOOLCHAIN_HAS_NEWLIB == 1
-    tags:
-      - lwm2m
-      - net
```
Comment (nashif) [inline tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: why is this not net related anymore? We have many platforms ignoring networking by tags, now all of thos will be building lwm2m for no good reason

**Review by nashif — CHANGES_REQUESTED** [review]

File: tests/net/lib/lwm2m/content_plain_text/testcase.yaml (hunk 1 above)
Comment (SeppoTakalo) [reply tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: I don't think there is any reason to filter out the `net` tagged tests on platforms like `qemu_cortex_m3` that is capable of running Network tests and we already provide configuration for it in LwM2M sample.
Seems pretty useless to filter away one of our mainly used test platform.

But because that `net` filtering exist, I have to workaround it.
None of these tests actually require networking, sockets are stubbed away here. But I added filtering so that it runs only on simulated platforms so we should not waste HW resources.
Is this not sufficient enough?

Can you suggest alternative? How can I run network related test on `qemu_cortex_m3` platform?

Comment (nashif) [reply tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: why do you need to run on qemu_cortex_m3? it is already covered by other emulation platforms

Comment (SeppoTakalo) [reply tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: One native X86 compiler (native_posix), one Cortex-M. I think it makes sense.

If not qemu_cortex_m3 what would you suggest?

Comment (PerMac) [reply tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: @nashif What do you mean by "now all of thos will be building lwm2m for no good reason"? I am referring to a fact that just because it can be build it doesn't mean it will be. As pointed in my previous comment, `platform_key` significantly limits the resulting builds. I thought that was the reason behind introduction of `platform_key`? (cc @teburd) Like I wrote, even with `--all` in the CLI the chosen scenario was built only on 4 boards.
I also just checked that adding platform_key benefits our on-target setup where we have ~10 different nrf boards. Without the change a single test is repeated on 3 boards (which has net capability). With the PR build is limited to a single one.

Comment (nashif) [reply tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: > @nashif What do you mean by "now all of thos will be building lwm2m for no good reason"? I am referring to a fact that just because it can be build it doesn't mean it will be.

with this PR:

```
gale:zephyr(lwm2m_refactor_tests): ./scripts/twister -p qemu_xtensa -T tests/net/lib/lwm2m/ -v
Renaming output directory to /home/nashif/zephyrproject/zephyr/twister-out.1
INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.4.0-247-g1b1ce125b39e
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 112
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
INFO    -  1/12 qemu_xtensa               tests/net/lib/lwm2m/lwm2m_engine/net.lwm2m.lwm2m_engine PASSED (qemu 2.075s)
INFO    -  2/12 qemu_xtensa               tests/net/lib/lwm2m/lwm2m_rd_client/net.lwm2m.lwm2m_rd_client_dtls PASSED (qemu 2.273s)
INFO    -  3/12 qemu_xtensa               tests/net/lib/lwm2m/lwm2m_rd_client/net.lwm2m.lwm2m_rd_client PASSED (qemu 2.275s)
INFO    -  4/12 qemu_xtensa               tests/net/lib/lwm2m/content_json/net.lwm2m.content_json PASSED (qemu 2.118s)
INFO    -  5/12 qemu_xtensa               tests/net/lib/lwm2m/content_raw_cbor/net.lwm2m.content_raw_cbor PASSED (qemu 2.082s)
INFO    -  6/12 qemu_xtensa               tests/net/lib/lwm2m/block_transfer/net.lwm2m.block_transfer PASSED (qemu 2.106s)
INFO    -  7/12 qemu_xtensa               tests/net/lib/lwm2m/lwm2m_registry/net.lwm2m.lwm2m_registry PASSED (qemu 2.061s)
INFO    -  8/12 qemu_xtensa               tests/net/lib/lwm2m/content_oma_tlv/net.lwm2m.content_oma_tlv PASSED (qemu 2.082s)
INFO    -  9/12 qemu_xtensa               tests/net/lib/lwm2m/content_link_format/net.lwm2m.content_link_format PASSED (qemu 2.061s)
INFO    - 10/12 qemu_xtensa               tests/net/lib/lwm2m/content_plain_text/net.lwm2m.content_plain_text PASSED (qemu 2.078s)
INFO    - 11/12 qemu_xtensa               tests/net/lib/lwm2m/engine/net.lwm2m.engine        PASSED (qemu 2.068s)
INFO    - 12/12 qemu_xtensa               tests/net/lib/lwm2m/content_senml_cbor/net.lwm2m.content_senml_cbor PASSED (qemu 2.129s)

INFO    - 12 test scenarios (12 test instances) selected, 0 configurations skipped (0 by static filter, 0 at runtime).
INFO    - 12 of 12 test configurations passed (100.00%), 0 failed, 0 errored, 0 skipped with 0 warnings in 18.89 seconds
INFO    - In total 233 test cases were executed, 0 skipped on 1 out of total 575 platforms (0.17%)
INFO    - 12 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

Without this PR:
```
gale:zephyr(main): ./scripts/twister -p qemu_xtensa -T tests/net/lib/lwm2m/ -v
Renaming output directory to /home/nashif/zephyrproject/zephyr/twister-out.2
INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.4.0-275-g71b522cfdd01
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/testplan.json
INFO    - JOBS: 112
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue

INFO    - 12 test scenarios (12 test instances) selected, 12 configurations skipped (12 by static filter, 0 at runtime).
INFO    - 0 of 12 test configurations passed (0.00%), 0 failed, 0 errored, 12 skipped with 0 warnings in 2.08 seconds
INFO    - In total 0 test cases were executed, 233 skipped on 1 out of total 575 platforms (0.17%)
INFO    - 0 test configurations executed on platforms, 0 test configurations were only built.
INFO    - Saving reports...
INFO    - Writing JSON report /home/nashif/zephyrproject/zephyr/twister-out/twister.json
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /home/nashif/zephyrproject/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```


Removing the `net` tag seems to be a workaround to run tests on a platform we have not used for running network tests.

> As pointed in my previous comment, platform_key significantly limits the resulting builds. 

Where exactly did I say anything about `platform_key`?

Comment (nashif) [reply tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: > If not qemu_cortex_m3 what would you suggest?

The same qemu we use for all other network based tests, i.e. qemu_x86, which also has userspace support, so it covers more than just the basics. 

> Edit: I would prefer to use same compiler `arm-zephyr-eabi` than our target HW, but without wasting HW resources.

isn't lwm2m and network in general hardware agnostic?

If you were testing some hardware features, sure, but testing networking does not really care what compiler and hardware you use AFAIK.

Comment (SeppoTakalo) [reply tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: Can you run the same with `native_posix` and without this PR?

I really don't see what is the point of limiting tests on platforms where they work just fine. The integration platforms are listed. According to documentation this should be the field used to limit the CI load.

Comment (SeppoTakalo) [reply tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: > Removing the net tag seems to be a workaround to run tests on a platform we have not used for running network tests.

Sure, but why do you block anyone from running module tests on those platforms? And what `arm-zephyr-eabi` platforms do we use for running network related tests?

**Comment by SeppoTakalo** [comment]
Please point me a filter set that runs these tests only on `native_posix`

That should be all that is required, however the documentation discourages the use of `platform_allow` field for that purpose.

File: tests/net/lib/lwm2m/content_plain_text/testcase.yaml (hunk 1 above)
Comment (nashif) [reply tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: > Sure, but why do you block anyone from running module tests on those platforms?

time, money and common sense.

If you want to run those test on a specific platform and you have a good reason for that, we should change the platform not to ignore the tag, but do not remove the tag from the tests, because the lwm2m is part of networking and has always been, so do not end up building the same code aimlessly on each platform and in each PR if it can be covered by one.


> And what `arm-zephyr-eabi` platforms do we use for running network related tests?

The question is if such tests need to be built and run with each compiler, I do not think so. This should be hardware agnostic at least on the emulation side. I am sure you get to build and run those on the target hardware as well outside of project CI, so you have the coverage for ARM or any other architecture as well.

Comment (SeppoTakalo) [reply tests/net/lib/lwm2m/content_plain_text/testcase.yaml:8]: I'm trying to suggest a sensible filtering so that this does not build on all.
And it is not sensible to assume that compiler architectures have differencies. So if Arm is our target HW, why not build tests with `arm-zephyr-eabi`. After all, that is installed on all developers on my organization.

Then bulk of testing I would prefer to use native_posix just because it is fastest and behaves nicely on GDB on local machine.

But what do you suggest? Why qemu_x86? Why install a toolchain just to run tests on architecture that matches neither the target HW or local development machine.

**Comment by SeppoTakalo** [comment]
I put back the `net` tag and dropped the `qemu_cortex_m3` platform away from integration platform.
I hope this is satisfactory.

**Review by PerMac — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Review by nashif — APPROVED** [review]
